ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/state/PivotBlockRetriever.java

```diff
@@ -0,0 +1,130 @@
+import tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncError;
+import tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncException;
+import tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncState;
+import tech.pegasys.pantheon.ethereum.eth.sync.tasks.GetPivotBlockHeaderFromPeerTask;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PivotBlockRetriever<C> {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int MAX_PIVOT_BLOCK_RETRIES = 3;
+  private final long pivotBlockNumber;
+  private final EthContext ethContext;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final Map<BlockHeader, AtomicInteger> confirmationsByBlockNumber =
+      new ConcurrentHashMap<>();
+  private final CompletableFuture<FastSyncState> result = new CompletableFuture<>();
+  private final Collection<GetPivotBlockHeaderFromPeerTask> getHeaderTasks =
+      new ConcurrentLinkedQueue<>();
+
+  public PivotBlockRetriever(
+      final ProtocolSchedule<C> protocolSchedule,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer,
+      final long pivotBlockNumber) {
+    this.pivotBlockNumber = pivotBlockNumber;
+    this.ethContext = ethContext;
+    this.ethTasksTimer = ethTasksTimer;
+    this.protocolSchedule = protocolSchedule;
+  }
+
+  @SuppressWarnings("rawtypes")
+  public CompletableFuture<FastSyncState> downloadPivotBlockHeader() {
+    final CompletableFuture[] requestFutures = requestHeaderFromAllPeers();
+
+    CompletableFuture.allOf(requestFutures)
+        .thenRun(
+            () -> {
+              // All requests have completed but we still haven't reached agreement on a header.
+              result.completeExceptionally(
+                  new FastSyncException(FastSyncError.PIVOT_BLOCK_HEADER_MISMATCH));
+            });
+    return result;
+  }
+
+  @SuppressWarnings("rawtypes")
+  private CompletableFuture[] requestHeaderFromAllPeers() {
+    final List<EthPeer> peersToQuery =
+        ethContext
+            .getEthPeers()
+            .availablePeers()
+            .filter(peer -> peer.chainState().getEstimatedHeight() >= pivotBlockNumber)
+            .collect(Collectors.toList());
+
+    final int confirmationsRequired = peersToQuery.size() / 2 + 1;
```

Comment:
I think we might want to have a cap on how many peers we want to query for this.  For example, if we have 50 peers, do we really need 25 confirmations?

Reply:
I think we want to be as certain as possible that it's the right node before fast syncing - it's a big point of weakness because we're trusting that the pivot block we select is correct, we can't perform any real checks that it actually is right.  I was actually wondering if we should abort if any of our peers returned a header that didn't match and maybe have a minimum number of peers that confirm the block (currently we might have 10 peers connected but 9 of them haven't reached the pivot block yet).